pkg/metadata/image_metadata.go

```diff
@@ -0,0 +1,156 @@
+// TODO(random-liu): Figure out a way to simplify this.
+// TODO(random-liu): Handle versioning
+
+// imageMetadataVersion is current version of image metadata.
+const imageMetadataVersion = "v1" // nolint
+
+// versionedImageMetadata is the internal struct representing the versioned
+// image metadata
+// nolint
+type versionedImageMetadata struct {
+	// Version indicates the version of the versioned image metadata.
+	Version string `json:"version,omitempty"`
+	ImageMetadata
+}
+
+// ImageMetadata is the unversioned image metadata.
+type ImageMetadata struct {
+	// Id of the image. Normally the Digest
+	ID string `json:"id,omitempty"`
+	// Other names by which this image is known.
+	RepoTags []string `json:"repo_tags,omitempty"`
+	// Digests by which this image is known.
+	RepoDigests []string `json:"repo_digests,omitempty"`
+	// Size of the image in bytes. Must be > 0.
+	Size uint64 `json:"size,omitempty"`
+	// Authentication configuration for pulling the image.
+	Auth *runtime.AuthConfig `json:"auth,omitempty"`
+	// CreatedAt is the created timestamp.
+	CreatedAt int64 `json:"created_at,omitempty"`
+}
+
+// ImageMetadataUpdateFunc is the function used to update ImageMetadata.
+type ImageMetadataUpdateFunc func(ImageMetadata) (ImageMetadata, error)
+
+// imageMetadataToStoreUpdateFunc generates a metadata store UpdateFunc from ImageMetadataUpdateFunc.
+func imageMetadataToStoreUpdateFunc(u ImageMetadataUpdateFunc) store.UpdateFunc {
+	return func(data []byte) ([]byte, error) {
+		meta := &ImageMetadata{}
+		if err := json.Unmarshal(data, meta); err != nil {
+			return nil, err
+		}
+		newMeta, err := u(*meta)
+		if err != nil {
+			return nil, err
+		}
+		return json.Marshal(newMeta)
+	}
+}
+
+// ImageMetadataStore is the store for metadata of all images.
+type ImageMetadataStore interface {
+	// Create creates an image's metadata from ImageMetadata in the store.
+	Create(ImageMetadata) error
+	// Get gets the specified image metadata.
+	Get(string) (*ImageMetadata, error)
+	// Update updates a specified image metatdata.
+	Update(string, ImageMetadataUpdateFunc) error
+	// List lists all image metadatas.
+	List() ([]*ImageMetadata, error)
+	// Delete deletes the image's metatdata from the store.
+	Delete(string) error
+}
+
+// imageMetadataStore is an implmentation of ImageMetadataStore.
+type imageMetadataStore struct {
+	store store.MetadataStore
+}
+
+// NewImageMetadataStore creates an ImageMetadataStore from a basic MetadataStore.
+func NewImageMetadataStore(store store.MetadataStore) ImageMetadataStore {
+	return &imageMetadataStore{store: store}
+}
+
+// Create creates a image's metadata from ImageMetadata in the store.
+func (s *imageMetadataStore) Create(metadata ImageMetadata) error {
+	data, err := json.Marshal(&metadata)
+	if err != nil {
+		return err
+	}
+	return s.store.Create(metadata.ID, data)
+}
+
+// Get gets the specified image metadata.
+func (s *imageMetadataStore) Get(digest string) (*ImageMetadata, error) {
```

Comment:
so yeah if there's a way to eliminate the need for us to store metadata for images we're all for it!